File: common/src/main/java/com/pluxity/facility/strategy/FloorService.java
package com.pluxity.facility.strategy;

import com.pluxity.facility.Facility;
import com.pluxity.facility.floor.Floor;
import com.pluxity.facility.floor.FloorRepository;
import com.pluxity.facility.floor.dto.FloorRequest;
import com.pluxity.facility.floor.dto.FloorResponse;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.CollectionUtils;

@Service
@RequiredArgsConstructor
public class FloorService {

    private final FloorRepository repository;

    @Transactional
    public <T extends Facility> void save(T facility, List<FloorRequest> floorRequests) {
        if (CollectionUtils.isEmpty(floorRequests)) {
            return;
        }

        List<Floor> floors =
                floorRequests.stream().map(request -> toEntity(facility, request)).toList();

        repository.saveAll(floors);
    }

    @Transactional
    public <T extends Facility> void update(T facility, List<FloorRequest> floorRequests) {
        repository.deleteByFacility(facility);

        save(facility, floorRequests);
    }

    @Transactional(readOnly = true)
    public <T extends Facility> List<FloorResponse> findAllByFacility(T facility) {
        return repository.findAllByFacility(facility).stream().map(FloorResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public <T extends Facility> Map<Facility, List<Floor>> findAllByFacilities(List<T> facilities) {
        if (CollectionUtils.isEmpty(facilities)) {
            return Collections.emptyMap();
        }
        return repository.findAllByFacilities(facilities).stream()
                .collect(Collectors.groupingBy(Floor::getFacility));
    }

    @Transactional
    public <T extends Facility> void delete(T facility) {
        repository.deleteByFacility(facility);
    }

    private Floor toEntity(Facility facility, FloorRequest request) {
        return Floor.builder()
                .facility(facility)
                .floorId(request.floorId())
                .name(request.name())
                .build();
    }
}

File: common/src/main/java/com/pluxity/facility/strategy/FloorService.kt
package com.pluxity.facility.strategy

import com.pluxity.facility.Facility
import com.pluxity.facility.floor.Floor
import com.pluxity.facility.floor.FloorRepository
import com.pluxity.facility.floor.dto.FloorRequest
import com.pluxity.facility.floor.dto.FloorResponse
import com.pluxity.facility.floor.dto.toFloorResponse
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.CollectionUtils

@Service
class FloorService(
    private val repository: FloorRepository,
) {
    @Transactional
    fun <T : Facility> save(
        facility: T,
        floorRequests: List<FloorRequest>?,
    ) {
        if (!floorRequests.isNullOrEmpty()) {
            val floors = floorRequests.map { request -> toEntity(facility, request) }
            repository.saveAll(floors)
        }
    }

    @Transactional
    fun <T : Facility> update(
        facility: T,
        floorRequests: List<FloorRequest>?,
    ) {
        repository.deleteByFacility(facility)
        save(facility, floorRequests)
    }

    @Transactional(readOnly = true)
    fun <T : Facility> findAllByFacility(facility: T): List<FloorResponse> =
        repository.findAllByFacility(facility).map { it.toFloorResponse() }

    @Transactional(readOnly = true)
    fun <T : Facility> findAllByFacilities(facilities: List<T>): Map<Facility, List<FloorResponse>> {
        if (CollectionUtils.isEmpty(facilities)) {
            return emptyMap()
        }
        return repository
            .findAllByFacilities(facilities)
            .groupBy({ it.facility!! }, { it.toFloorResponse() })
    }

    @Transactional
    fun <T : Facility> delete(facility: T) {
        repository.deleteByFacility(facility)
    }

    private fun toEntity(
        facility: Facility,
        request: FloorRequest,
    ): Floor =
        Floor(
            facility = facility,
            floorId = request.floorId,
            name = request.name,
        )
}
